DEPRECATION.md

```diff
@@ -0,0 +1,91 @@
+# Deprecation Policy Proposal
+
+As AllenNLP matures we will inevitably need to deprecate old models, interfaces and features. This document provides guidelines that attempt to minimize churn. Reducing churn helps users build next generation models with tools they've already mastered.
+
+### Terminology note
+
+We _deprecate_ some component to communicate that a preferred alternative exists. A _deprecated_ component does not necessarily have to be _removed_ from the library.
+
+### Goals
+
+1. Promote backward compatibility where not onerous.
+2. Describe code-level and release process for both deprecation and removal.
+
+### Non-goals
+
+1. Eliminate all deprecation and removal.
+2. Legislate. Good judgment by AllenNLP developers is much preferred.
+
+## Decision Flow
+
+Say you've found some old code that needs some care. You have a fix in mind, maybe already a PR. Great, let's get it merged! But let's give some thought to backward compatibility along the way.
+
+1. Determine whether backward compatibility is even an issue.
+
+  * __Likely safe__
+
+    * New code
+    * Internal code _TODO: How is this determined?_
+    * Implementation change only
+
+  * __Care required__
+
+    * File/class/function names
+      * Including names passed to `Registrable.register`!
+    * Function signatures
+    * Jsonnet config
+    * Model internals that affect shape of saved weights
+
+2. If so, first consider how critical the change is.
+
+  * __Non-critical__
+
+    * Minor name changes
+    * "Cleaner" APIs with no functional benefit
+
+  * __Medium__
+
+    * Name/API that is actively confusing multiple users.
+      * This means Github issues, messages on user channels, etc.
+    * Useful new feature
+
+  * __High__
+
+    * Major bugs
+
+3. Non-critical changes we should happily avoid. Otherwise, let's try to make the change in a backward compatible manner.
+
+  * Options include:
+    * When deprecating, say, a class, leave a shim to its replacement.
+    * A function signature can likely be changed with a keyword argument.
+    * You can decorate with `Registrable.register` multiple times.
+    * Features can be hidden behind flags.
+    * Files can simply be forked. (for extreme cases)
+
+  * Likely no deprecation needs to occur in this case. However, in the spirit of TOOWTDI, one may still mark the old component as deprecated while _not_ removing it. Usages in AllenNLP should be removed so that our warnings are not spammy. Though silencing warnings is an option, we should "eat our own dog food" or we can hardly expect our users to migrate.
+
```

Comment:
@schmmd , also I'm not sure I understand your point (delivered offline) about not emitting deprecation warnings. I do currently see them when I run pytest. Didn't we discuss this as one issue with softly deprecating features?

Reply:
@brendan-ai2 we wont have any warnings once https://github.com/allenai/allennlp/pull/2282 merges--and no deprecations should introduce any warnings.

Reply:
I read the discussion for https://github.com/allenai/allennlp/pull/2282 and that's not how I interpreted it. It seems like we absolutely want to warn when we use a deprecated component. Otherwise how are we going to track down the usages?

For instance:
https://github.com/allenai/allennlp/pull/2282#issuecomment-452048702
https://github.com/allenai/allennlp/pull/2282#discussion_r245492233